feat: adapted bidrectional meshPolicy documentation (#482)
meshcloud/meshcloud-docs: docs/meshcloud.policies.md

## Introduction: what are meshPolicies?

> A meshPolicy describes a restriction between two meshSubjects. meshSubjects are [meshObjects](meshcloud.index.md#introduction) that are globally defined by your organization but not the other way around. This means only selected meshObjects are meshSubjects.

Author: I want to make clear that we call it in context of meshPolicies meshSubjects. For me it is also fine to describe this behavios like you've mentioned.

> **meshCustomer**, **meshProject**, **meshUser/Group** and **meshLandingZone** are meshSubjects

You can find the authoritative meshSubject on the left side and the affected meshSubject on the right side. Each meshPolicy describes which [tag](meshcloud.metadata-tags.md) values have to be present on both meshSubjects, in order to comply with the meshPolicy. To decide if a meshPolicy is compliant it is necessary to select an evaluation strategy how you want to evaluate the authoritative meshSubject against the affected meshSubject. meshPolicies are enforced in various places when attaching meshSubjects. How meshPolicies can be configured is described [here](administration.mesh-policies.md).

> meshPolicies will only be enforced for meshCustomers. meshPartners are excluded.

> ### Authoritative and affected meshSubject

Reviewer: d: I'm not sure about the terms authoritative meshSubject and affected meshSubject since we never mentioned it in the panel. But for comprehension I can understand why it can be useful. 
But on the other hand we only have two authoritative meshSubject saying meshCustomer and meshProject. So that another possibility could be to use them right away.  Saying not calling the section "Authoritative and affected meshSubject" but having for example this section:

**What can meshPolicies enforce?**

meshPolicies can enforce that
- meshCustomer restricts the meshProjects 
- meshCustomer restricts new assignments meshUsers/Groups
- meshProject restricts new assignments meshUsers/Groups
- meshProject restricts new assignments meshLandingZones 

based on the defined tag in the meshPolicies.

So just an idea. Also fine "authoritative meshSubject and affected meshSubject" section :)

Author: I prefer the direct naming because in the future we can have other auth meshSubjects besides the meshCustomer and meshProject.

The terms authoritative and affected meshSubject shall make clear in which direction the meshPolicy will be evaluated. In general the authoritative meshSubject restricts the affected meshSubject depending on the selected evaluation strategy.

> We've a limited selection of meshSubject combinations. Valid combinations are

Author: Fixed.

- meshCustomer -> meshUser/Group
- meshCustomer -> meshProject
- meshProject -> meshUser/Group

> - meshProject -> meshLandingZone

Author: I prefer the list view.

> ### meshPolicy evaluation strategy

Reviewer 2: This section is perfect to link to from the comment above in the admin area.

Author: Fxied.

A meshPolicy evaluation stategy describes a strategy how authoritative and affected meshSubjects shall be evaluated in context of a meshPolicy. There exists two different selectable evaluation options `Subset` and `Intersection`.

> #### Subset

Reviewer 2: It would be great to have an example of this. It is too theoretical at the moment.

Author: We'll improve the docs afterwards. This is not the final state.

Describes an evaluation strategy that the tag values of the affected meshSubject **must be a subset** of the authoritative meshSubject tag values.

> > Recommended for

Reviewer: i: maybe you can go for Recommended for meshPolicy between:

Author: Fixed.

>
> - meshCustomer -> meshProject

> #### Intersection

Reviewer 2: And also an example here.

Author: Same here.

## Introduction: what are meshPolicies?

> A meshPolicy describes a restriction between two [meshObjects](meshcloud.index.md#introduction) that are globally defined by your organization. We've everytime the authoritative meshObject on the left side and the affected meshObject on the right side. The relationship will be only compliant if the affected meshObject values are a subset of the authoritative meshObject values. Each meshPolicy describes which [tag](meshcloud.metadata-tags.md) values have to be present on both meshObjects, in order to comply with the meshPolicy. meshPolicies are enforced in various places when attaching meshObjects. How meshPolicies can be configured is described [here](administration.mesh-policies.md).

Reviewer 3: ```suggestion
A meshPolicy describes a restriction between two [meshObjects](meshcloud.index.md#introduction) that are globally defined by your organization. You can find the authoritative meshObject on the left side and the affected meshObject on the right side. The relationship will only be compliant if the affected meshObject values are a subset of the authoritative meshObject values. Each meshPolicy describes which [tag](meshcloud.metadata-tags.md) values have to be present on both meshObjects, in order to comply with the meshPolicy. meshPolicies are enforced in various places when attaching meshObjects. How meshPolicies can be configured is described [here](administration.mesh-policies.md).
```
I guess to mention the subset is no longer true right? Also I think users will struggle with authoritative meshObject. "Attaching meshObject" sounds very abstract. Can you write create and edit instead and provide examples.

Author: I've updated the docs. Please take second look.

1. Creating a new project

>     1. When adding a meshPlatform with a meshLandingZone to a new project, all policies are evaluated between a 'meshProject' and a 'meshLandingZone'

Reviewer 3: ```suggestion
    1. When adding a meshPlatform with a meshLandingZone to a new project, all policies are evaluated between a 'meshProject' and a 'meshLandingZone'
```
Can we make this a bullet list instead of a numbered list. I dont think the order matters

Author: Sure

### How are meshPolicies evaluated

> As mentioned before, meshPolicies are built on top of meshStack's [tagging](./meshcloud.metadata-tags.md) system. This is done by evaluating two meshObjects on their tags. There is the "direction" between the affected meshObject and authoritative meshObject in a meshPolicy: the meshObjects are evaluated unidirectionally.

Reviewer 3: ```suggestion
As mentioned before, meshPolicies are built on top of meshStack's [tagging](./meshcloud.metadata-tags.md) system. This is done by evaluating two meshObjects on their tags. There is the "direction" between the affected meshObject and authoritative meshObject in a meshPolicy: the meshObjects are evaluated unidirectionally.
```
This part of the documentation needs to be rewritten when we change the policy dialog in the meshPanel

> If a meshPolicy is defined, the two meshObjects' tags will be evaluated on a so-called set intersection. All affected meshObject tag values must be present within the authoritative meshObject tag values. Otherwise the meshPolicy is not compliant. If neither side have the given tag, the meshPolicy is also successful.

Reviewer 3: With the newest change around at least on and subset I think this part is not correct anymore. Also we would need a second table

Author: Fixed.

Take this meshPolicy violation as an example (see the picture below), where we have a meshPolicy defined on 'meshCustomer' and 'meshProject' with the evaluation strategy 'Subset', both on the `environment` tag.

![Example Policy Error Message](assets/mesh_policies/policies_example_error_message.png)

The project that is being created called 'my-example-project-prod' wants environment `prod` and the meshCustomer 'managed-customer' the project is being created in, has environments `dev`, `test` and `qa` defined. This means there is a mismatch as `<prod>` is not inside `<dev, test, qa>`. In order to solve this problem, we have to pick an environment that is defined on the meshCustomer, e.g. `dev`. After picking a valid environment value, we can save the project again, and (if we don't violate any further policies) the project is successfully created, including the right compliance for your organization! ✅

### Are there any other places where policies are enforced?

> Besides end-users being impacted at the places above, there are also other places where meshPolicy violations could be caused. The diagram below describes all possible relationships between meshSubjects and the behavior that is expected, depending on who is doing the change.

Author: Added mermaid graph.

> Besides end-users being impacted at the places above, there are also other places where meshPolicy violations could be caused. The diagram below describes all possible relationships between meshSubjects and the behavior that is expected, depending on who is doing the change.

Reviewer: ```
flowchart TB
    subgraph Legend
    start1["blocking violationon new assignments"]:::legendClass --> stop1[ ]:::legendClass
    start2["logging violation but not blocking"]:::legendClass -.-> stop2[ ]:::legendClass
    start3["blocks on creation and editing"]:::legendClass --> stop3[ ]:::legendClass
    end
    subgraph In case of policy violation
      meshCustomer:::authorativeClass --> meshProject
		  meshProject:::authorativeClass -.-> meshCustomer
		  meshProject --> meshLandingZone
		  meshLandingZone -.-> meshProject
		  meshUser/Group --> meshCustomer
		  meshCustomer -.-> meshUser/Group
		  meshUser/Group -.-> meshProject
		  meshProject --> meshUser/Group
    end

classDef authorativeClass fill:#dae8fc;
classDef legendClass height:0px;
linkStyle 2 stroke: blue;
linkStyle 3 stroke: blue;```

![image](https://user-images.githubusercontent.com/64793818/160790445-95c48160-e345-4249-8438-651da2bf0202.png)

> Besides end-users being impacted at the places above, there are also other places where meshPolicy violations could be caused. The diagram below describes all possible relationships between meshSubjects and the behavior that is expected, depending on who is doing the change.

Author: Added.

![Example Policy Error Message](assets/mesh_policies/policies_example_error_message.png)

The project that is being created called 'my-example-project-prod' wants environment `prod` and the meshCustomer 'managed-customer' the project is being created in, has environments `dev`, `test` and `qa` defined. This means there is a mismatch as `<prod>` is not inside `<dev, test, qa>`. In order to solve this problem, we have to pick an environment that is defined on the meshCustomer, e.g. `dev`. After picking a valid environment value, we can save the project again, and (if we don't violate any further policies) the project is successfully created, including the right compliance for your organization! ✅

### Are there any other places where policies are enforced?

Besides end-users being impacted at the places above, there are also other places where meshPolicy violations could be caused. The diagram below describes all possible relationships between meshSubjects and the behavior that is expected, depending on who is doing the change.

> ```mermaid

Reviewer 2: I find this diagram a little bit messy but I also do not really have any better proposals.

Isn't it the case though that the directions of these arrows have to do with the authoritative / affected mechanism? Maybe that explains it better than this diagram?

Author: We want to show whats the meaning is between the subjects independent of authoritative and affected.
We've decided within the proposal to use mermaid. So I try to implement the diagram depending on the notion guide https://www.notion.so/meshcloud/meshcloud-docs-diagrams-with-Mermaid-JS-c3deaea4ce8f40ad95fdde1a56b678e0 .

### How are meshPolicies evaluated

As mentioned before, meshPolicies are built on top of meshStack's [tagging](./meshcloud.metadata-tags.md) system. This is done by evaluating two meshSubjects on their tags. There is the "direction" between the affected meshSubject and authoritative meshSubject in a meshPolicy: the meshSubjects are evaluated unidirectionally.
If a meshPolicy is defined, the two meshSubjects' tags will be evaluated depending on the evaluation strategy. If neither side have the given tag, the meshPolicy is also successful.
To make this more clear, the table below describes certain conditions and their behavior when policies are applied on tags.

> In this example, we're looking at a meshPolicy between a meshProject's `environment` tag, and a meshCustomer's `environment` tag with the following allowed values: `dev`, `qa`, and `prod`. The meshProject is the affected meshSubject and the meshCustomer is the authoritative meshSubject.

Author: Fxied.

### How are meshPolicies evaluated

As mentioned before, meshPolicies are built on top of meshStack's [tagging](./meshcloud.metadata-tags.md) system. This is done by evaluating two meshSubjects on their tags. There is the "direction" between the affected meshSubject and authoritative meshSubject in a meshPolicy: the meshSubjects are evaluated unidirectionally.
If a meshPolicy is defined, the two meshSubjects' tags will be evaluated depending on the evaluation strategy. If neither side have the given tag, the meshPolicy is also successful.
To make this more clear, the table below describes certain conditions and their behavior when policies are applied on tags.

In this example, we're looking at a meshPolicy between a meshProject's `environment` tag, and a meshCustomer's `environment` tag with the following allowed values: `dev`, `qa`, and `prod`. The meshProject is the affected meshSubject and the meshCustomer is the authoritative meshSubject. The chosen evaluation strategy for this example is 'Subset'.

| meshProject  | meshCustomer | Result | Explanation |

> | ------------ | ------------ | ------ | ----------- |

Reviewer 2: I would have expected this table at the evaluation strategy part. I think it is really helpful

Author: I've added another table to the evaluation strategy section. This section makes sense because here we've the explicit meshCustomer meshProject example.
